cmd/forwarder/log.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2021 The forwarder Authors. All rights reserved.
+// Use of this source code is governed by a MIT
+// license that can be found in the LICENSE file.
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/saucelabs/sypl"
+	"github.com/saucelabs/sypl/fields"
+	"github.com/saucelabs/sypl/formatter"
+	"github.com/saucelabs/sypl/level"
+	"github.com/saucelabs/sypl/options"
+	"github.com/saucelabs/sypl/output"
+)
+
+const (
+	defaultTimeFormat = "2006-01-02T15-04-05"
+	infoLevel         = "info"
+	name              = "proxy"
+)
+
+// logConfig is a configuration for the logger.
+type logConfig struct {
+	FileLevel string `validate:"required,gte=3"`
+	FilePath  string `validate:"required"`
+	Level     string `validate:"required,gte=3"`
+}
+
+func defaultLogConfig() logConfig {
+	return logConfig{
+		FileLevel: infoLevel,
+		FilePath:  filepath.Join(os.TempDir(), fmt.Sprintf("%s-%s.log", name, time.Now().Format(defaultTimeFormat))),
+		Level:     infoLevel,
+	}
+}
+
+func newLogger(cfg logConfig) forwarderLogger {
+	l := sypl.NewDefault(name, level.MustFromString(cfg.Level))
+	l.AddOutputs(
+		output.File(cfg.FilePath, level.MustFromString(cfg.FileLevel)).SetFormatter(formatter.Text()),
+	)
+	l.PrintlnWithOptions(&options.Options{
+		Fields: fields.Fields{
+			"fileLevel": cfg.FileLevel,
+			"filePath":  cfg.FilePath,
+			"level":     cfg.Level,
+		},
+	}, level.Trace, "Logging is configured")
+	return forwarderLogger{l}
+}
+
+// forwarderLogger is a wrapper around sypl.Logger that implements forwarder.Logger.
+// It is needed because sypl.Logger methods return sypl.ISypl, which makes it impossible work with anything else...
+type forwarderLogger struct {
+	*sypl.Sypl
+}
+
+func (l forwarderLogger) Errorf(format string, args ...interface{}) {
```

Comment:
I'd suggest adding these methods to Sypl instead (if it doesn't already have those). This way you can define an external logger requirement as a standard logger. And use Sypl if it's needed. The point is to remove Sypl as a logger type in the `New` function.

Reply:
Changing Sypl would be a big mess, it's not about adding sth to it, it's about changing its interface.
All Sypl logging functions return the logger as `ISypl`, this includes the glorious `Breakpoint` function, all the printing interfaces and more.

```go
	// Error prints @ the Error level.
	Error(args ...interface{}) ISypl

	// Errorf prints according with the format @ the Error level.
	Errorf(format string, args ...interface{}) ISypl

	// Errorlnf prints according with the format @ the Error level, also adding
	// a new line to the end.
	Errorlnf(format string, args ...interface{}) ISypl

	// Errorln prints, also adding a new line to the end @ the Error level.
	Errorln(args ...interface{}) ISypl
```

https://github.com/saucelabs/sypl/blob/a11e44ec19587f758153915803c9fb175804cef5/interface.go#L115

> The point is to remove Sypl as a logger type in the New function.

Sypl is removed like you say. Only cli depends on `sypl` library but it also needs some glue code.
I can add a package `log/sypllog` that would expose the glue code so that it can be reused.

Reply:
>> Changing Sypl would be a big mess...

oh yeah, I didn't realize it.

> I can add a package log/sypllog that would expose the glue code so that it can be reused

Maybe add it to Sypl? My biggest concern is that we need to add the latest Forwarder to the existing Sauce Connect code and release a new binary sometime soon.

Reply:
I checked that, the sc uses go modules, it specifies to use `github.com/saucelabs/forwarder v0.3.4` precisely so none of the changes we are doing here would affect the release.